src/adaXT/random_forest/RandomForest.py

```diff
@@ -0,0 +1,355 @@
+            for tree in self.trees:
+                predictions.append(self._predict_proba_single_tree(tree))
+        else:
+            with Pool(self.n_jobs) as p:
+                predictions = p.map(
+                    self._predict_proba_single_tree, self.trees)
+
+        return predictions
+
+    def fit(self, X: np.ndarray, Y: np.ndarray):
+        """
+        Function used to fit a forrest using many DecisionTrees for the given data
+
+        Parameters
+        ----------
+        X : np.ndarray
+            feature values
+        Y : np.ndarray
+            response values
+        """
+        if not self.bootstrap and self.max_samples:
+            raise AttributeError(
+                "Bootstrap can not be False while max_samples is set")
+
+        X, Y = self.check_input(X, Y)
+        self.features = SharedNumpyArray(X)
+        self.outcomes = SharedNumpyArray(Y)
+        self.n_obs, self.n_features = self.features._shape
+        if self.forrest_type == "Classification":
+            self.classes = np.unique(self.outcomes.read())
+        if self.max_samples is None:
+            self.max_samples = self.n_obs
+        self.trees = [
+            DecisionTree(
+                tree_type=self.forrest_type,
+                criteria=self.criterion,
+                max_depth=self.max_depth,
+                impurity_tol=self.impurity_tol,
+                min_samples_split=self.min_samples_split,
+                min_samples_leaf=self.min_samples_leaf,
+                min_improvement=self.min_improvement,
+                max_features=self.max_features,
+                splitter=self.splitter) for _ in range(
+                self.n_estimators)]
+
+        # Fit trees
+        self.__build_trees()
+
+        # Register that the forrest was succesfully fitted
+        self.forest_fitted = True
+
+        return self
+
+    # Function that returns an ndarray of random ints used as the
+    # sample_indices
+    def __get_sample_indices(self):
+        if self.bootstrap:
+            return np.random.randint(
+                low=0, high=self.n_obs, size=self.max_samples)
+        else:
+            return None
+
+    def predict(self, X: np.ndarray):
+        """
+        Predicts a y-value for given X values using the trees of the forrest.
+
+        Parameters
+        ----------
+        X : np.ndarray
+            (N, M) numpy array with features to predict
+
+        Returns
+        -------
+        np.ndarray
+            (N) numpy array with the prediction
+        """
+        if not self.forest_fitted:
+            raise AttributeError(
+                "The forrest has not been fitted before trying to call predict")
+
+        self.predict_values = SharedNumpyArray(X)
+
+        # Predict using all the trees, each column is the predictions from one
+        # tree
+        tree_predictions = self.__predict_trees()
+        self.predict_values.unlink()
+
+        if self.forrest_type == "Regression":
```

Comment:
I wonder whats the best way to handle this so we don't need to doubly specify the prediction. Could it make sense to have an aggregation function, with some sane defaults (e.g., mean and majority vote). Currently I can't think of what else one would like to do.

Reply:
One could also consider adding the aggregation method into the decision tree predict class?

Reply:
We could add a parameter for the forest called aggregation_type with the signature: "func(List[Trees] ->  float) or {“mean”, “majority_vote”}" such that a user can specify a function of their own or use one of our defaults. What do you think about that? 

I don't think it would make sense to have the aggregation method on the DecisionTree predict class, as this would somehow seems to be in the scope of the Forest and not on a single tree.